### PR TITLE
Update monster looting

### PR DESCRIPTION
**Pull request by moseleyc**
## Description
-Add item looting, and few minor fixes

## Checklist
- [x] Tested and working locally
- [ ] Reviewed
- [ ] Tested and working on dev server
- [ ] Deployed to staging

## How to QA
- [x] Kill creature with one loot-table, item is dropped (based on chance value)
- [x] Kill creature with one loot-table, gold is dropped (based on chance value)
- [x] Kill creature with multiple loot-tables, item(s) are dropped (based on chance value & loot-table)
- [x] Kill creature with multiple loot-tables, gold(s) are dropped (based on chance value & loot-table)
- [x] Kill creature with no loot-table, gold is dropped (based on loot, not loot-table)

- [x] User kills creature, able to pick up items
- [x] User kills creature, otherUser unable to pick up items until monsterDeathPileTime duration exceeded
- [x] User kills creature, grpedUser able to pick up items

## Impacted Areas in Hybrasyl

- Creatures.Designer (Changed Rng variable from static to instance)
- hybrasylExtensions (Changed Rng variable from static to instance)
- Monolith (Added Loot Assignment to Spawned Mobs, Added SpawnLoot internal class to facilitate loot assignment)
- Creature (Added FirstHitter to handle mob tagging, Added mob tagging logic to Damage method)
- Monster (Added looting logic for Monsters.OnDeath to assign permitted looters for gold and items)
- User (Modified properties handling loot permissions)
- VisibleObject (Modified properties handling loot permissions, adjusted/updated CanBeLooted logic with new properties)
- Utilities (Added MonsterLoot timeout for non-permitted looters)
- World (Added LootSet XML loading, added a reset of item properties (for looting permissions) to avoid permissions continuing in perpetuity (which was happening without this))
- prop.ps1 (Added new folder for LootSets XML's)

### CODE REVIEW

File: HybrasylXML/XSD/hybrasylExtensions.cs [hunk 1]
```diff
@@ -312,6 +312,32 @@ namespace Hybrasyl.Creatures
     public partial class Spawn
     {
         protected static Random Rng = new Random();
```
Comment (baughj) [inline HybrasylXML/XSD/hybrasylExtensions.cs]: not your code, but we should probably change `Rng` here to be an instance variable

File: hybrasyl/Objects/Creature.cs [hunk 2]
```diff
@@ -854,6 +854,16 @@ public virtual bool UseCastable(Castable castObject, Creature target = null)
...
+                    if(this is User)
+                    {
+                        if (tar.MonsterDeathPileAllowedLooters.Count > 0 && !(tar.MonsterDeathPileAllowedLooters.Contains(Name))) continue;
+                        else if (tar.MonsterDeathPileAllowedLooters.Count == 0)
+                        {                            
+                            var theUser = this as User;
+                            if (theUser.Grouped) tar.MonsterDeathPileAllowedLooters = theUser.Group.Members.Select(user => user.Name).ToList();
+                            else tar.MonsterDeathPileAllowedLooters.Add(theUser.Name);
+                        }
+                    }
```
Comment (baughj) [inline hybrasyl/Objects/Creature.cs]: all of this should be moved to `OnDeath()`

File: hybrasyl/Objects/VisibleObject.cs [hunk 3]
```diff
@@ -46,6 +46,9 @@ public class VisibleObject : WorldObject
         public List<string> DeathPileAllowedLooters { get; set; }
         public DateTime? DeathPileTime { get; set; }
 
+        public List<string> MonsterDeathPileAllowedLooters { get; set; }
+        public DateTime? MonsterDeathPileTime { get; set; }
```
Comment (baughj) [inline hybrasyl/Objects/VisibleObject.cs]: I am not sure a differentiation here is needed (e.g. use `DeathPileAllowedLooters`).

File: hybrasyl/Objects/VisibleObject.cs [hunk 4]
```diff
@@ -78,13 +83,23 @@ public bool CanBeLooted(string username, out string error)
...
+                error = "These items do not belong to you.";
+            }
+            else
+            {
+                // Check if the item is part of a death pile
+                if (DeathPileTime == null) return true;
+                if (DeathPileOwner == username) return true;
+                if (DeathPileAllowedLooters.Contains(username) &&
+                    (DateTime.Now - DeathPileTime.Value).TotalSeconds > Constants.DEATHPILE_GROUP_TIMEOUT) return true;
+                if ((DateTime.Now - DeathPileTime.Value).TotalSeconds > Constants.DEATHPILE_RANDO_TIMEOUT) return true;
```
Comment (baughj) [inline hybrasyl/Objects/VisibleObject.cs]: Collapse all of these checks down to one set of if/else if imo

File: hybrasyl/Objects/Creature.cs (hunk 2 above)
Comment (moseleyc) [reply hybrasyl/Objects/Creature.cs]: I thought this might be applicable here to also handle attacking users, essentially the first attack "tagging"

Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: we have `LastHitter`; it might make sense for there to be `FirstHitter` to handle tagging. We may want to set that in `Damage()`. `OnDeath()` should still handle the allowed looters and should use `First/LastHitter` as needed.

**Comment by moseleyc** [comment]
New committ f349ee4 hasn't been fully QA'd yet, but based on a single test, everything appears to work now (including ImportLoot/LootSets).

**Comment by moseleyc** [comment]
Changes reviewed now.

**Comment by norrismiv** [comment]
There are merge conflicts on this branch. Please fix the conflicts before continuing the review.

**Comment by moseleyc** [comment]
Conflicts resolved.

**Comment by moseleyc** [comment]
Just in case this needed to be explicitly said, I've also done a QA check on the "merged" version.

File: HybrasylXML/XSD/hybrasylExtensions.cs [hunk 5]
```diff
@@ -311,7 +311,7 @@ namespace Hybrasyl.Creatures
 
     public partial class Spawn
     {
-        protected static Random Rng = new Random();
+        protected Random Rng;
```
Comment (baughj) [inline HybrasylXML/XSD/hybrasylExtensions.cs]: this needs to be initialized, either `Rng = new Random()` or in constructor

File: hybrasyl/Monolith.cs [hunk 6]
```diff
@@ -138,6 +143,155 @@ private static void SpawnMonster(Monster monster, Map map)
...
+        /// <summary>
+        /// Calculate the total amount of lootable gold based on the base amount plus any additional from the loot table(s).
+        /// </summary>
+        /// <returns>Gold uint</returns>
+        public uint LootableGold()
+        {
+            uint lootableGold = 0;
+            var spawnLootGold = _spawn.Loot.Gold;
+
+            if (spawnLootGold != null)
```
Comment (baughj) [inline hybrasyl/Monolith.cs]: `SpawnLootGold` can never be null (it is always initialized - check `Loot.Designer.cs` line 87), so a lot of this stuff can be expressed more concisely. I pushed an update to your branch to enforce some defaults and update XML structure a bit.

File: hybrasyl/Monolith.cs [hunk 7]
```diff
@@ -138,6 +143,155 @@ private static void SpawnMonster(Monster monster, Map map)
...
+            {
+                lootableGold += ((uint)_rng.Next((int)(spawnLootGold?.Min ?? 0), (int)(spawnLootGold?.Max ?? 0)));
+            }
+            if (_spawn.Loot.Table != null || _spawn.Loot.Set != null)
+            {
+                _spawnLootTable.ForEach(lootTable => lootableGold += ((uint)_rng.Next((int)(lootTable.Gold?.Min ?? 0), (int)(lootTable.Gold?.Max ?? 0))));
```
Comment (baughj) [inline hybrasyl/Monolith.cs]: This seems very strange to me. Why are you checking if a different data structure is null (_spawn.Loot.Table) and then using ForEach on a pregenerated (_spawnLootTable) one? Just use `foreach` on the output from CreateSpawnLootTable. In the event that there is no table or sets, the list should be returned as empty.

File: hybrasyl/Monolith.cs [hunk 8]
```diff
@@ -138,6 +143,155 @@ private static void SpawnMonster(Monster monster, Map map)
...
+            List<ItemObject> lootableItems = new List<ItemObject>();
+            var worldItems = Game.World.WorldData.Values<Hybrasyl.Items.Item>();
+
+            if (_spawnLootTable.Count > 0)
+            {
+                foreach (var table in _spawnLootTable)
+                {
+                    foreach (var item in table.Items.Items)
+                    {
+                        for (int i = 0; i < table.Items.Rolls; i++)
```
Comment (baughj) [inline hybrasyl/Monolith.cs]: I think there may be a little misunderstanding here. You don't roll on each item; you roll to get a random item from the _table itself_. So the roll/chance doesn't impact each item; it impacts the entire table.

For instance, consider the following fragment:

```xml
<Items Rolls="6" Chance="0.25">
  <Item>Stick</Item>
  <Item>Mold</Item>
  <Item>Hot Garbage</Item>
  <Item>Swonko's Letter</Item>
</Items>
```

The intent is that this would result in 6 rolls being made, with a 25% chance to get an item off of the table (per roll). The item selection is random, _unless_ the drop is enforced using `Always`:

`<Item Always="true">Stick</Item>` - would always result in _one_ stick dropping

Also, you need to account for `Unique` (only drop once, regardless of # of rolls / # of hits) and `Variants`. 

`Variants` in particular need to be handled in the first pass to handle woodlands areas.

File: hybrasyl/Objects/Creature.cs [hunk 9]
```diff
@@ -801,6 +815,14 @@ public virtual void RegenerateMp(double mp, Creature regenerator = null)
 
         public virtual void Damage(double damage, Enums.Element element = Enums.Element.None, Enums.DamageType damageType = Enums.DamageType.Direct, Castables.DamageFlags damageFlags = Castables.DamageFlags.None, Creature attacker = null, bool onDeath = true)
         {
+            var attackerUser = attacker as User;
+            if(attackerUser != null)
```
Comment (baughj) [inline hybrasyl/Objects/Creature.cs]: This is a little hard to follow, I think the logic can be simplified. All `Creatures` have the `Id` and `Name` property so, I don't think there is a need to check if it can be converted to a user, or to convert them to user; we just need to know it _is_ a user. The last part seems to be enforcing a "tagging" check, which can be done at the main scope. The check  for `ActiveUsersByName` is also incorrect here since I believe you mean to check whether the existing `FirstHitter` is online - not the attacker (presumably the latter is always true. :P)

Similarly I think there is a logic error in the tagging check (we want to make sure the `FirstHitter`'s group contains `attackerUser` - right now it checks if `attackerUser`'s group contains `attackerUser` which, again, will always be true.

To sum up:

```csharp
if ((attacker is User && FirstHitter == null) || (FirstHitter != null && !Game.World.ActiveUsersByName.ContainsKey(FirstHitter.Name))
  _mFirstHitter = attacker.Id;

if (FirstHitter != attacker && (FirstHitter as User).Grouped && (FirstHitter as User).Group.Members.Contains(attacker.Name))
  return;


 
```

File: hybrasyl/Objects/VisibleObject.cs [hunk 10]
```diff
@@ -86,13 +87,15 @@ public bool CanBeLooted(string username, out string error)
...
-            // Check if the item is part of a death pile
-            if (DeathPileTime == null) return true;
+
+            // Check if the item is part of a death pile or recent monster dropped items
+            if (ItemDropTime == null) return true;
             if (DeathPileOwner == username) return true;
-            if (DeathPileAllowedLooters.Contains(username) &&
-                (DateTime.Now - DeathPileTime.Value).Seconds > Constants.DEATHPILE_GROUP_TIMEOUT) return true;
-            if ((DateTime.Now - DeathPileTime.Value).Seconds > Constants.DEATHPILE_RANDO_TIMEOUT) return true;
+            if (ItemDropAllowedLooters.Contains(username) && (DateTime.Now - ItemDropTime.Value).TotalSeconds > Constants.DEATHPILE_GROUP_TIMEOUT) return true;
```
Comment (baughj) [inline hybrasyl/Objects/VisibleObject.cs]: I think you mean `||` here.

File: hybrasyl/Objects/VisibleObject.cs [hunk 11]
```diff
@@ -86,13 +87,15 @@ public bool CanBeLooted(string username, out string error)
...
+            if ((!DeathPileOwner.Equals(string.Empty)) && (DateTime.Now - ItemDropTime.Value).TotalSeconds > Constants.DEATHPILE_RANDO_TIMEOUT) return true;
```
Comment (baughj) [inline hybrasyl/Objects/VisibleObject.cs]: No need for Equals here, either use `== string.Empty` or `string.IsNullOrEmpty(DeathPileOwner)`

File: hybrasyl/Objects/VisibleObject.cs [hunk 12]
```diff
@@ -86,13 +87,15 @@ public bool CanBeLooted(string username, out string error)
...
+            if (DeathPileOwner.Equals(string.Empty) && (DateTime.Now - ItemDropTime.Value).TotalSeconds > Constants.MONSTER_LOOT_DROP_RANDO_TIMEOUT) return true;
```
Comment (baughj) [inline hybrasyl/Objects/VisibleObject.cs]: I would check the expiration of timeouts first and then the values of strings - since the timeout overrules everything else. I believe it would make the logic easier to follow, as well.

File: prep.ps1 [hunk 13]
```diff
@@ -18,6 +18,7 @@ New-Item -force -ItemType directory -Path $mydocuments\Hybrasyl\world\xml\creatu
 New-Item -force -ItemType directory -Path $mydocuments\Hybrasyl\world\xml\items
 New-Item -force -ItemType directory -Path $mydocuments\Hybrasyl\world\xml\itemvariants
 New-Item -force -ItemType directory -Path $mydocuments\Hybrasyl\world\xml\localization
+New-Item -Force -ItemType directory -Path $mydocuments\Hybrasyl\world\xml\lootsets
```
Comment (baughj) [inline prep.ps1]: Nice job remembering to update this. Can you update the `example.zip` file too?

Comment (norrismiv) [reply prep.ps1]: I thought this was to be done via the initial configuration when launching the server? I advise against pushing powershell scripts expecting the users to know how to use them.

Comment (norrismiv) [reply prep.ps1]: Example:

```if(!Directory.Exists(MapFileDirectory))
{
  Directory.Creat(MapFileDirectory);
}```

File: HybrasylXML/XSD/hybrasylExtensions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (moseleyc) [reply HybrasylXML/XSD/hybrasylExtensions.cs]: This is initialized in the Constructor (Creatures.Designer.cs line 59).

File: prep.ps1 (hunk 13 above)
Comment (baughj) [reply prep.ps1]: Currently it isn't; we say to use the PS1 or unzip the zipfile in the readme. This is to make it easier to start the server for new users.

File: HybrasylXML/XSD/hybrasylExtensions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (baughj) [reply HybrasylXML/XSD/hybrasylExtensions.cs]: Yep, OK. Sorry, didn't see it.

File: prep.ps1 (hunk 13 above)
Comment (norrismiv) [reply prep.ps1]: That's my point.. this should be fixed on server initialization rather than using the powershell script.

File: HybrasylXML/XSD/hybrasylExtensions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (norrismiv) [reply HybrasylXML/XSD/hybrasylExtensions.cs]: Creatures.Designer.cs is a generated file and should not be hand editted.

File: HybrasylXML/XSD/Creatures.Designer.cs [hunk 14]
```diff
@@ -56,6 +56,7 @@ public Spawn()
             this._damage = new Damage();
             this._respawn = new Respawn();
             this._variance = ((float)(1F));
+            Rng = new Random();
```
Comment (norrismiv) [inline HybrasylXML/XSD/Creatures.Designer.cs]: This cannot occur here unless this is included in the xsd.

File: prep.ps1 (hunk 13 above)
Comment (baughj) [reply prep.ps1]: OK - out of scope here as the PS1 went in via one of my PRs. So we should open a github issue to fix.

File: HybrasylXML/XSD/hybrasylExtensions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (moseleyc) [reply HybrasylXML/XSD/hybrasylExtensions.cs]: Should I return this to the original `protected static Random Rng = new Random();` ?

File: prep.ps1 (hunk 13 above)
Comment (norrismiv) [reply prep.ps1]: ok that's fine, then that ps file should be removed from the scope of loot fixes then and added as a separate PR.

..or it can just be brought into scope here.

File: HybrasylXML/XSD/hybrasylExtensions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (norrismiv) [reply HybrasylXML/XSD/hybrasylExtensions.cs]: that's probably the best bet. the designer files will be overwritten any time an xsd change is made.

you can add it to the hybrasylExtensions class though, and extend it through there if needed.

[the remark this answers is not in the document]
Comment (baughj) [reply HybrasylXML/XSD/hybrasylExtensions.cs]: yes, please put in the extensions, and make it not `static`

[the remark this answers is not in the document]
Comment (moseleyc) [reply HybrasylXML/XSD/hybrasylExtensions.cs]: Done.

File: prep.ps1 (hunk 13 above)
Comment (moseleyc) [reply prep.ps1]: Github Ticket Created: #139

File: hybrasyl/Monolith.cs (hunk 7 above)
Comment (moseleyc) [reply hybrasyl/Monolith.cs]: Will remove. This was part of a different design, so as noted isn't needed now.

File: hybrasyl/Monolith.cs (hunk 6 above)
Comment (moseleyc) [reply hybrasyl/Monolith.cs]: Will update this.

File: hybrasyl/Objects/Creature.cs (hunk 9 above)
Comment (moseleyc) [reply hybrasyl/Objects/Creature.cs]: Yea, I was over thinking this. Essentially, I want the "tagging" to be reset if all the users leave the game or the map to avoid having creatures "tagged" indefinitely. I'll have to think how to do that separately.

Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: I think a timeout for this case is more useful rather than trying to keep data around to handle the case indefinitely. I suggest 5 minutes?

Comment (moseleyc) [reply hybrasyl/Objects/Creature.cs]: Okay. Do you want me to create a constant for that timeout (so we can make it a defined field in the config xml or where ever these are going to go?)

Comment (moseleyc) [reply hybrasyl/Objects/Creature.cs]: Logic adjusted.

File: hybrasyl/Monolith.cs (hunk 8 above)
Comment (moseleyc) [reply hybrasyl/Monolith.cs]: 1) Do I have the implementation of the tables correct then, with regard to the rolls? Or should it be possible for the same table to be used potentially multiple times (up to the number of rolls)?

2) Using your example, if Rolls = 1 and `<Item Always="true">Stick</Item>`, the only the stick would ever drop, correct?

3) What exactly are `variants`? Given that they only require a string value, I didn't understand what this was meant to be, so I was going to do this in a second pass.

Comment (Caeldeth) [reply hybrasyl/Monolith.cs]: Re: 3

Variants are item variants, such as "Deoch Jade Ring", "Good Battle Sword", "Might Mythril Bracer", etc.  Am uncertain how they are constructed, but defined via xml.

Comment (moseleyc) [reply hybrasyl/Monolith.cs]: Of course. Thanks for clarifying. That part makes sense to me now.

Comment (moseleyc) [reply hybrasyl/Monolith.cs]: Do I need to consider the `variants` of the item itself as well? Or just the `variants` referred to in the loot table? My assumption is the latter, but wanted to confirm.

Comment (baughj) [reply hybrasyl/Monolith.cs]: Sorry it took a while for me to respond, was indisposed yesterday.

1) Yes, it is possible for the same table to be used multiple times.
2) "Always" was intended to always be *additive* and only occurs once (so for 4 items plus an always item, you would get the always item and a randomly selected item from the 4).
3) Variants are referred to in the loot table. e.g. stick, variant great would have a chance to drop Great Stick.

File: hybrasyl/Monolith.cs [hunk 15]
```diff
@@ -138,6 +143,155 @@ private static void SpawnMonster(Monster monster, Map map)
...
+            }
+            return spawnLootTable;
+        }
+
+        /// <summary>
+        /// Used to Convert Hybrasyl.LootTable to Creatures.LootTable
+        /// </summary>
+        /// <param name="lootTableToConvert">Hybrasyl.LootTable to Convert</param>
+        /// <returns>Creatures.LootTable</returns>
+        private Creatures.LootTable ConvertLootTable(Hybrasyl.LootTable lootTableToConvert)
```
[the remark this answers is not in the document]
Comment (moseleyc) [reply hybrasyl/Monolith.cs]: If I've understood correctly, this is done.

File: hybrasyl/Monolith.cs (hunk 8 above)
Comment (moseleyc) [reply hybrasyl/Monolith.cs]: This is done.

**Comment by moseleyc** [comment]
QA'd, should be good to go now.

File: hybrasyl/Objects/Creature.cs [hunk 16]
```diff
@@ -801,6 +805,14 @@ public virtual void RegenerateMp(double mp, Creature regenerator = null)
 
         public virtual void Damage(double damage, Enums.Element element = Enums.Element.None, Enums.DamageType damageType = Enums.DamageType.Direct, Castables.DamageFlags damageFlags = Castables.DamageFlags.None, Creature attacker = null, bool onDeath = true)
         {
+            if (attacker is User)
+            {
+                if (FirstHitter == null || !Game.World.ActiveUsersByName.ContainsKey(FirstHitter.Name) || ((DateTime.Now - LastHitTime).TotalSeconds > Constants.MONSTER_TAGGING_TIMEOUT)) FirstHitter = attacker;
+                if (attacker != FirstHitter && !((FirstHitter as User).Group?.Members.Contains(attacker) ?? false)) return;
```
Comment (baughj) [inline hybrasyl/Objects/Creature.cs]: It occurs to me this logic will have unexpected results for certain edge cases (pvp, for one). For now, what do you think about updating it to `if (attacker is User && this is Monster)?`

File: hybrasyl/Objects/VisibleObject.cs [hunk 17]
```diff
@@ -50,19 +50,21 @@ public class VisibleObject : WorldObject
         public string DisplayText { get; set; }
 
         public string DeathPileOwner { get; set; }
-        public List<string> DeathPileAllowedLooters { get; set; }
-        public DateTime? DeathPileTime { get; set; }
+        public List<string> ItemDropAllowedLooters { get; set; }
+        public DateTime? ItemDropTime { get; set; }
+        public ItemDropType ItemDropType { get; set; }
```
Comment (baughj) [inline hybrasyl/Objects/VisibleObject.cs]: These can stay here for now but a future refactor is to clean up this and move it to `Item` instead as it is nonsensical for, say, a `Creature` to have a drop type, or to have allowed looters (since these properties are in `VisibleObject`)

File: hybrasyl/Objects/Creature.cs (hunk 16 above)
Comment (moseleyc) [reply hybrasyl/Objects/Creature.cs]: Makes sense to me. Will add now.

File: hybrasyl/Objects/VisibleObject.cs (hunk 17 above)
Comment (moseleyc) [reply hybrasyl/Objects/VisibleObject.cs]: Okay. I'll add that to my list.

Would that also apply to DeathPileOwner?

Comment (baughj) [reply hybrasyl/Objects/VisibleObject.cs]: Yes.
